core/image/jpeg/jpeg.odin

```diff
@@ -0,0 +1,927 @@
+				if img.channels == 0 && img.depth == 0 && img.width == 0 && img.height == 0 {
+					return img, .Encountered_SOS_Before_SOF
+				}
+
+				if .do_not_decompress_image in options {
+					return img, nil
+				}
+
+				// Length
+				compress.read_data(ctx, u16be) or_return
+				num_components := compress.read_u8(ctx) or_return
+				if num_components != 1 && num_components != 3 {
+					return img, .Invalid_Number_Of_Channels
+				}
+
+				for _ in 0..<num_components {
+					component_id := cast(Component)compress.read_u8(ctx) or_return
+					if component_id < .Y || component_id > .Cr {
+						return img, .Image_Does_Not_Adhere_to_Spec
+					}
+
+					// high 4 is DC, low 4 is AC
+					coefficient_indices := compress.read_u8(ctx) or_return
+					dc_table_idx := coefficient_indices >> 4
+					ac_table_idx := coefficient_indices & 0xF
+
+					if (dc_table_idx < 0 || dc_table_idx > 3) || (ac_table_idx < 0 || ac_table_idx > 3) {
+						return img, .Invalid_Huffman_Table_Index
+					}
+
+					color_components[component_id].dc_table_idx = dc_table_idx
+					color_components[component_id].ac_table_idx = ac_table_idx
+				}
+				// TODO: These aren't used for sequential DCT, only progressive and lossless.
+				Ss := compress.read_u8(ctx) or_return
+				_ = Ss
+				Se := compress.read_u8(ctx) or_return
+				_ = Se
+				Ah_Al := compress.read_u8(ctx) or_return
+				_ = Ah_Al
+
+				blocks = make([]Block, block_height * block_width)
+
+				previous_dc: [Component]i16
+
+				restart_interval *= color_components[.Y].v_sampling_factor * color_components[.Y].h_sampling_factor
+				#no_bounds_check for y := 0; y < mcu_height; y += color_components[.Y].v_sampling_factor {
+					for x := 0; x < mcu_width; x += color_components[.Y].h_sampling_factor {
+						blk := y * block_width + x
+
+						if restart_interval != 0 && blk % restart_interval == 0 {
+							previous_dc[.Y] = 0
+							previous_dc[.Cb] = 0
+							previous_dc[.Cr] = 0
+							byte_align(ctx)
+						}
+						for c in 1..=img.channels {
+							c := cast(Component)c
+							for v in 0..<color_components[c].v_sampling_factor {
+							h_loop:
+								for h in 0..<color_components[c].h_sampling_factor {
+									mcu := &blocks[(y + v) * block_width + (h + x)][c]
+									dc_table := huffman[.DC][color_components[c].dc_table_idx]
+									ac_table := huffman[.AC][color_components[c].ac_table_idx]
+									quantization_table := quantization[color_components[c].quantization_table_idx]
+
+									length := get_symbol(ctx, dc_table)
+
+									if length > 11 {
+										fmt.println("DC coefficient length greater than 11")
+									}
+
+									dc_coeff := cast(i16)read_bits_msb(ctx, length)
+
+									if length != 0 && dc_coeff < (1 << (length - 1)) {
+										dc_coeff -= (1 << length) - 1
+									}
+									mcu[0] = (dc_coeff + previous_dc[c]) * cast(i16)quantization_table[0]
+									previous_dc[c] = dc_coeff + previous_dc[c]
+
+									for i := 1; i < COEFFICIENT_COUNT; i += 1 {
```

Comment:
Use range based for loop `for i in 1..<COEFFICIENT_COUNT`

Reply:
I can't use a range-based for loop because I increment `i` inside the loop